### ISSUE TITLE
Length of authorization Subject property causes "key was too long" on MySQL 5.6

### ISSUE THREAD

**Issue by codeaid** [openiddict/openiddict-core#719]
I'm currently working on a project that will be hosted on AWS Aurora Serverless, which unfortunately only supports MySQL 5.6 and after having implemented OpenIddict and generated migrations I am getting the following error when it gets to genrating indexes for the authorization model:

```
[20:29:41 Error] Microsoft.EntityFrameworkCore.Database.Command
Failed executing DbCommand (4ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
CREATE INDEX `IX_oidc_authorizations_ApplicationId_Status_Subject_Type` ON `oidc_authorizations` (`ApplicationId`, `Status`, `Subject`, `Type`);
MySql.Data.MySqlClient.MySqlException (0x80004005): Specified key was too long; max key length is 767 bytes ---> MySql.Data.MySqlClient.MySqlException (0x80004005): Specified key was too long; max key length is 767 bytes
   at MySqlConnector.Core.ResultSet.ReadResultSetHeaderAsync(IOBehavior ioBehavior) in C:\projects\mysqlconnector\src\MySqlConnector\Core\ResultSet.cs:line 42
   at MySql.Data.MySqlClient.MySqlDataReader.ActivateResultSet(ResultSet resultSet) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 80
   at MySql.Data.MySqlClient.MySqlDataReader.ReadFirstResultSetAsync(IOBehavior ioBehavior) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 302
   at MySql.Data.MySqlClient.MySqlDataReader.CreateAsync(MySqlCommand command, CommandBehavior behavior, ResultSetProtocol resultSetProtocol, IOBehavior ioBehavior) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 287
   at MySqlConnector.Core.TextCommandExecutor.ExecuteReaderAsync(String commandText, MySqlParameterCollection parameterCollection, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\Core\TextCommandExecutor.cs:line 37
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQueryAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlCommand.cs:line 261
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQuery() in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlCommand.cs:line 62
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues)

MySql.Data.MySqlClient.MySqlException (0x80004005): Specified key was too long; max key length is 767 bytes ---> MySql.Data.MySqlClient.MySqlException (0x80004005): Specified key was too long; max key length is 767 bytes
   at MySqlConnector.Core.ResultSet.ReadResultSetHeaderAsync(IOBehavior ioBehavior) in C:\projects\mysqlconnector\src\MySqlConnector\Core\ResultSet.cs:line 42
   at MySql.Data.MySqlClient.MySqlDataReader.ActivateResultSet(ResultSet resultSet) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 80
   at MySql.Data.MySqlClient.MySqlDataReader.ReadFirstResultSetAsync(IOBehavior ioBehavior) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 302
   at MySql.Data.MySqlClient.MySqlDataReader.CreateAsync(MySqlCommand command, CommandBehavior behavior, ResultSetProtocol resultSetProtocol, IOBehavior ioBehavior) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 287
   at MySqlConnector.Core.TextCommandExecutor.ExecuteReaderAsync(String commandText, MySqlParameterCollection parameterCollection, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\Core\TextCommandExecutor.cs:line 37
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQueryAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlCommand.cs:line 261
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQuery() in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlCommand.cs:line 62
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteNonQuery(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationCommand.ExecuteNonQuery(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationCommandExecutor.ExecuteNonQuery(IEnumerable`1 migrationCommands, IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.UpdateDatabase(String targetMigration, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabase.<>c__DisplayClass0_1.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
Specified key was too long; max key length is 767 bytes
```

I see that the `OpenIddictAuthorizationConfiguration` class is setting maximum length of the `Subject` property to 450, which I'm guessing is causing issues. I know I can override the length myself by calling `HasMaxLength(something else)` but I don't know what the possible maximum size is. Besides it would be nice if this feature would work out of the box without custom tweaks.

Here's what my startup looks like (cleaned out of irrelevant code):

```csharp
namespace SomeApp
{
    public class Startup
    {
        public void ConfigureServices(IServiceCollection services)
        {
            ConfigureDb(services);
            ConfigureIdentity(services);
            ConfigureOpenIdDict(services);
        }

        private void ConfigureDb(IServiceCollection services)
        {
            var connectionString = Configuration.GetConnectionString("Database");
            var migrationsAssembly = typeof(ApplicationDbContext).Assembly.GetName().Name;

            services.AddDbContextPool<ApplicationDbContext>(options =>
            {
                options.UseMySql(connectionString,
                    mysqlOptions =>
                    {
                        mysqlOptions.MigrationsAssembly(migrationsAssembly);
                        mysqlOptions.ServerVersion(new Version(5, 6), ServerType.MySql);
                    });

                options.UseOpenIddict<int>();
            });
        }

        private void ConfigureIdentity(IServiceCollection services)
        {
            services.AddIdentity<ApplicationUser, ApplicationRole>()
                .AddEntityFrameworkStores<ApplicationDbContext>()
                .AddDefaultTokenProviders();
        }

        private void ConfigureOpenIdDict(IServiceCollection services)
        {
            services.AddOpenIddict()
                .AddCore(options =>
                {
                    options.UseEntityFrameworkCore()
                        .UseDbContext<ApplicationDbContext>()
                        .ReplaceDefaultEntities<int>();
                });
        }
    }
}
```

And here's my database context:

```csharp
namespace SomeApp
{
    public class ApplicationDbContext : IdentityDbContext<ApplicationUser, ApplicationRole, int>
    {
        public ApplicationDbContext(DbContextOptions<ApplicationDbContext> options) : base(options)
        {
        }

        protected override void OnModelCreating(ModelBuilder builder)
        {
            base.OnModelCreating(builder);

            ConfigureIdentityTables(builder);
            ConfigureIdentityFields(builder);

            ConfigureOpenIddictTables(builder);
        }

        private void ConfigureIdentityTables(ModelBuilder builder)
        {
            builder.Entity<ApplicationRole>().ToTable("roles");
            builder.Entity<IdentityRoleClaim<int>>().ToTable("roles_claims");

            builder.Entity<ApplicationUser>().ToTable("users");
            builder.Entity<IdentityUserClaim<int>>().ToTable("users_claims");
            builder.Entity<IdentityUserLogin<int>>().ToTable("users_logins");
            builder.Entity<IdentityUserRole<int>>().ToTable("users_roles");
            builder.Entity<IdentityUserToken<int>>().ToTable("users_tokens");
        }

        private void ConfigureIdentityFields(ModelBuilder builder)
        {
            builder.Entity<ApplicationRole>(entity =>
            {
                entity.Property(model => model.Name).HasMaxLength(50);
                entity.Property(model => model.NormalizedName).HasMaxLength(50);
            });

            builder.Entity<ApplicationUser>(entity =>
            {
                entity.Property(model => model.Email).HasMaxLength(150);
                entity.Property(model => model.NormalizedEmail).HasMaxLength(150);
                entity.Property(model => model.UserName).HasMaxLength(150);
                entity.Property(model => model.NormalizedUserName).HasMaxLength(150);
            });
        }

        private void ConfigureOpenIddictTables(ModelBuilder builder)
        {
            builder.UseOpenIddict<int>();

            builder.Entity<OpenIddictApplication<int>>().ToTable("oidc_applications");
            builder.Entity<OpenIddictAuthorization<int>>().ToTable("oidc_authorizations");
            builder.Entity<OpenIddictScope<int>>().ToTable("oidc_scopes");
            builder.Entity<OpenIddictToken<int>>().ToTable("oidc_tokens");
        }
    }
}
```

I've gone through the examples a couple times and it looks like I'm not missing anything configuration wise. I actually had OpenIddict setup previously for a different project that was using MySQL 5.7+ where it worked with no issues and can't figure out what I'm doing wrong this time (apart from the version).

**Comment by codeaid** [comment]
My relevant dependencies:

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">
  <PropertyGroup>
    <TargetFramework>netcoreapp2.2</TargetFramework>
    <StartupObject>SomeApp.Program</StartupObject>
    <LangVersion>7.3</LangVersion>
  </PropertyGroup>
  <ItemGroup>
    </PackageReference>
    <PackageReference Include="Microsoft.EntityFrameworkCore">
      <Version>2.2.1</Version>
    </PackageReference>
    <PackageReference Include="OpenIddict" Version="2.0.0" />
    <PackageReference Include="OpenIddict.EntityFrameworkCore">
      <Version>2.0.0</Version>
    </PackageReference>
    <PackageReference Include="Pomelo.EntityFrameworkCore.MySql">
      <Version>2.1.4</Version>
    </PackageReference>
  </ItemGroup>
</Project>
```

Edit:
Adding the following code to my `ConfigureOpenIddictTables` method fixes the issue:

```csharp
builder.Entity<OpenIddictAuthorization<int>>(entity =>
{
    entity.Property(model => model.Subject).HasMaxLength(250);
});
builder.Entity<OpenIddictToken<int>>(entity =>
{
    entity.Property(model => model.Subject).HasMaxLength(250);
});
```

Reason for having `OpenIddictToken` there is because once authorization gets fixed it fails with the same error when migrating tokens.

Still, not perfect as seeing how it's set to be 450 by deafult makes me wonder if 250 is enough to store the subjects.

**Comment by kevinchalet** [comment]
I tested locally with MySQL 8.0.12 and I was unable to reproduce it so yeah, the fact Aurora uses an older version is likely the root cause here.

FWIW, setting an explicit maximum length was precisely made to avoid issues with MySQL but I should probably have tested with older versions.

The dilemma is that the `Subject` column corresponds to arbitrary data - the `sub` claim added to the claims principal when creating the authentication ticket, which is something set by the developer or the library that comes on top of OpenIddict when using one (e.g the OrchardCore OpenID module).

When using ASP.NET Core Identity, it will mostly depend on the user property you use as the `sub` claim. If you've configured Identity to do `options.ClaimsIdentity.UserIdClaimType = OpenIddictConstants.Claims.Subject` like in the samples, then the `sub` claim will be the user identifier. 

By default, Identity uses "GUID as strings" primary keys (like OpenIddict), so the user identifier/subject shouldn't exceed 36 ASCII characters.

TL;DR, in your case, 250 is a very reasonable max length. We may want to adopt the same one directly in OpenIddict to avoid future issues with older versions of MySQL. Let me know if you're interested in fixing that.
